components/smfSupport.ts

```diff
@@ -0,0 +1,182 @@
+                    )
+                ) {
+                    return
+                }
+
+                return parse(
+                    readFileSync(
+                        join(
+                            process.env.LOCALAPPDATA,
+                            "Simple Mod Framework",
+                            "blobs",
+                            lastServerSideData.blobs[name] ||
+                                lastServerSideData.blobs[name.slice(1)],
+                        ),
+                    ).toString(),
+                )
+            },
+        )
+    }
+
+    private handleContracts(lastServerSideData: LastServerSideData) {
+        if (!lastServerSideData?.contracts) return
+
+        for (const contractData of Object.values(
+            lastServerSideData.contracts,
+        )) {
+            this.controller.addMission(contractData)
+
+            if (contractData.SMF?.destinations?.addToDestinations) {
+                if (
+                    contractData.SMF.destinations.peacockIntegration !== false
+                ) {
+                    this.handleDestination(contractData)
+                }
+            }
+        }
+    }
+
+    private handleDestination(contractData: MissionManifest) {
+        const location = contractData.Metadata.Location
+        const id = contractData.Metadata.Id
+        const placeBefore = contractData.SMF.destinations.placeBefore
+        const placeAfter = contractData.SMF.destinations.placeAfter
+
+        if (placeBefore) {
+            const index =
+                this.controller.missionsInLocations[location].indexOf(
+                    placeBefore,
+                )
+            this.controller.missionsInLocations[location].splice(index, 0, id)
+        } else if (placeAfter) {
+            const index =
+                this.controller.missionsInLocations[location].indexOf(
+                    placeAfter,
+                ) + 1
+            this.controller.missionsInLocations[location].splice(index, 0, id)
+        } else {
+            this.controller.missionsInLocations[location].push(id)
+        }
+    }
+
+    private handleUnlockables(lastServerSideData: LastServerSideData) {
+        if (lastServerSideData?.unlockables) {
+            this.controller.configManager.configs["allunlockables"] =
+                lastServerSideData.unlockables.slice(1)
+        }
+    }
+
+    public async initSMFSupport(modFrameworkDataPath: string) {
+        if (!(modFrameworkDataPath && existsSync(modFrameworkDataPath))) {
+            return
+        }
+
+        log(
+            LogLevel.INFO,
+            "Simple Mod Framework installed - using the data it outputs.",
+            "boot",
+        )
+
+        const lastServerSideData = this.lastDeploy?.lastServerSideStates
+
+        this.handleUnlockables(lastServerSideData)
+        this.handleContracts(lastServerSideData)
+        this.handleBlobs(lastServerSideData)
+
+        if (lastServerSideData?.peacockPlugins) {
```

Comment:
The elvis operator here is not technically necessary since `lastServerSideStates` will always be defined if the framework has written something to AppData.

Reply:
The type can't guarantee that I think.

Reply:
That's fine then